src/Std/Data/DHashMap/Internal/RawLemmas.lean

```diff
@@ -830,15 +833,526 @@ theorem contains_keys [EquivBEq α] [LawfulHashable α] (h : m.1.WF) {k : α} :
+
+theorem getKey_constInsertMany_list_of_mem [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)}
+    {k k' : α} (k_beq : k == k')
+    (distinct : l.Pairwise (fun a b => (a.1 == b.1) = false))
+    (mem : k ∈ (l.map Prod.fst))
+    {h'} :
+    (insertMany m l).1.getKey k' h' = k := by
+  simp_to_model using getKey_insertListConst_of_mem
+
+theorem getKey!_constInsertMany_list_of_contains_eq_false [EquivBEq α] [LawfulHashable α] [Inhabited α]
+    (h : m.1.WF) {l : List (α × β)} {k : α}
+    (h' : (l.map Prod.fst).contains k = false) :
+    (insertMany m l).1.getKey! k = m.getKey! k := by
+  simp_to_model using getKey!_insertListConst_of_contains_eq_false
+
+theorem getKey!_constInsertMany_list_of_mem [EquivBEq α] [LawfulHashable α] [Inhabited α] (h : m.1.WF)
+    {l : List (α × β)}
+    {k k' : α} (k_beq : k == k')
+    (distinct : l.Pairwise (fun a b => (a.1 == b.1) = false))
+    (mem : k ∈ (l.map Prod.fst)) :
+    (insertMany m l).1.getKey! k' = k := by
+  simp_to_model using getKey!_insertListConst_of_mem
+
+theorem getKeyD_constInsertMany_list_of_contains_eq_false [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)} {k fallback : α}
+    (h' : (l.map Prod.fst).contains k = false) :
+    (insertMany m l).1.getKeyD k fallback = m.getKeyD k fallback := by
+  simp_to_model using getKeyD_insertListConst_of_contains_eq_false
+
+theorem getKeyD_constInsertMany_list_of_mem [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)}
+    {k k' fallback : α} (k_beq : k == k')
+    (distinct : l.Pairwise (fun a b => (a.1 == b.1) = false))
+    (mem : k ∈ (l.map Prod.fst)) :
+    (insertMany m l).1.getKeyD k' fallback = k := by
+  simp_to_model using getKeyD_insertListConst_of_mem
+
+theorem size_constInsertMany_list [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)}
+    {distinct : l.Pairwise (fun a b => (a.1 == b.1) = false)} :
+    (∀ (a : α), m.contains a -> (l.map Prod.fst).contains a = false) →
+    (insertMany m l).1.1.size = m.1.size + l.length := by
+  simp_to_model using length_insertListConst
+
+theorem isEmpty_constInsertMany_list_eq_false_of_isEmpty_eq_false [EquivBEq α] [LawfulHashable α]
+    (h : m.1.WF) {l : List (α × β)} :
+    (m.1.isEmpty = false) → (insertMany m l).1.1.isEmpty = false := by
+  simp_to_model using isEmpty_insertListConst_eq_false_of_isEmpty_eq_false
+
+theorem isEmpty_constInsertMany_list [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)} :
+    (insertMany m l).1.1.isEmpty ↔ m.1.isEmpty ∧ l.isEmpty := by
+  simp_to_model using isEmpty_insertListConst
+
+theorem get?_constInsertMany_list_of_contains_eq_false [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    {l : List (α × β)} {k : α}
+    (h' : (l.map Prod.fst).contains k = false) :
+    get? (insertMany m l).1 k = get? m k := by
+  simp_to_model using getValue?_insertListConst_of_contains_eq_false
+
+theorem get?_constInsertMany_list_of_mem [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
```

Comment:
I guess strictly speaking you're missing a fourth variant of this lemma, relating the (non-const) `get?` to `Const.insertMany`. On the other hand, I'm now starting to think that it would have been easier to just provide `get?_insertMany_list_of_mem` and `Const.get?_constInsertMany_list_of_mem` and have the user use `Const.get?_eq_get?` to rewrite their goal to match one of these. Then we could rename `Const.get?_constInsertMany_list_of_mem` back to `Const.get?_insertMany_list_of_mem` which is  simpler.

I'm not decided what the best approach is here. What do you think?

Reply:
I'm inclined towards only keeping the `Const.get?_constInsertMany_list_of_mem` variants (and renaming them).